ADBench/plot_graphs.py

```diff
@@ -459,32 +459,85 @@ def generate_graph(figure_info, sorted_vals_by_tool):
     if not do_show:
         pyplot.close(figure)
 
+def get_plot_data(all_graphs):
+    '''Creates a plot data from the files, produced by the global runner.'''
+
+    plot_data = []
+
+    for graph_info in all_graphs:
+        (graph, function_type) = graph_info
+        build_type = graph[0]
+        objective = graph[1]
+        maybe_test_size = graph[2] if len(graph) == 3 else ""
+
+        print(f"\n    Build type: {build_type}\n"
+              f"    Function type: {function_type}\n"
+              f"    Objective: {objective}\n"
+              f"    Test size: {maybe_test_size}")
+        sorted_vals_by_tool = get_sorted_vals_by_tool(objective, graph, function_type)
+
+        plot_data.append({
+            "build": build_type,
+            "objective": objective,
+            "function_type": function_type,
+            "test_size": maybe_test_size,
+            "values": [
+                {
+                    "tool": tool_name,
+                    "time": time_vals,
+                    "variable_count": var_count_vals,
+                    "violations": violation_vals
+                }
+                for tool_name, var_count_vals, time_vals, violation_vals in sorted_vals_by_tool
+            ]
+        })
+
+    return plot_data
+
+def extract_vals_and_figure_info_from_plot_data(data):
+    '''Extracts vals sorted by the tool and the figure info from the plot
+    data for the single figure.'''
+
+    sorted_vals_by_tool = [
+        (val["tool"], val["variable_count"], val["time"], val["violations"])
+        for val in data["values"]
+    ]
+
+    figure_info = namedtuple(
+        "figure_info",
+        "idx, build_type, objective, maybe_test_size, function_type"
+    )
+
+    figure_info.build_type = data["build"]
+    figure_info.objective = data["objective"]
+    figure_info.maybe_test_size = data["test_size"]
+    figure_info.function_type = data["function_type"]
+
+    return sorted_vals_by_tool, figure_info
+
 def main():
     print_messages()
 
-    # Loop through each of graphs to be created
-    for (figure_idx, t) in enumerate(all_graphs, start=1):
-        figure_info = namedtuple(
-            "figure_info",
-            "idx, build_type, objective, maybe_test_size, function_type"
-        )
+    print("\nGetting plot data...\n")
+    plot_data = get_plot_data(all_graphs)
 
-        (graph, figure_info.function_type) = t
-        figure_info.build_type = graph[0]
-        figure_info.objective = graph[1]
-        figure_info.maybe_test_size = graph[2] if len(graph) == 3 else ""
+    # Loop through each of graphs to be created
+    print("\nGenerating graphs...\n")
+    for (figure_idx, data) in enumerate(plot_data, start=1):
+        sorted_vals_by_tool, figure_info = extract_vals_and_figure_info_from_plot_data(data)
```

Comment:
It seems to me that `figure_idx` should be passed into `extract_vals_and_figure_info_from_plot_data` and set within that function, rather than being set afterwards.  What do you think?

Reply:
Actually, I think, it would be even better if `idx` were not a part of figure info at all. It seems like the best option

Reply:
Changed

Reply:
Do you mean instead to pass it in directly as an argument to `generate_graph`?  If so then I agree.  That sounds like an even better idea.

Reply:
Yes, that's what I meant